### PR TITLE
KTOR-2579 Add file based http cache storage for client

### PR DESCRIPTION
**Pull request by rsinukov**
https://youtrack.jetbrains.com/issue/KTOR-2579

Everything with suffix `Legacy` is moving of old storage related things to new files.

### CODE REVIEW

File: ktor-client/ktor-client-core/jvm/src/io/ktor/client/plugins/cache/storage/FileCacheStorage.kt [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    }
+
+    private fun key(url: Url) = hex(url.toString().encodeToByteArray())
+
+    private suspend fun writeCache(urlHex: String, caches: List<CachedResponseData>) = coroutineScope {
+        val mutex = mutexes.computeIfAbsent(urlHex) { Mutex() }
+        mutex.withLock {
+            val channel = ByteChannel()
+            File(directory, urlHex).outputStream().buffered().use { output ->
+                launch {
```
Comment (e5l) [inline ktor-client/ktor-client-core/jvm/src/io/ktor/client/plugins/cache/storage/FileCacheStorage.kt:50]: The file could be closed before the launch is finished if the `launch` block suspends. Could you please add a test with a large body and move the file `use` inside?

**Review by e5l — CHANGES_REQUESTED** [review]
lgtm, please check the launch block

File: ktor-client/ktor-client-core/jvm/src/io/ktor/client/plugins/cache/storage/FileCacheStorage.kt (hunk 1 above)
Comment (rsinukov) [reply ktor-client/ktor-client-core/jvm/src/io/ktor/client/plugins/cache/storage/FileCacheStorage.kt:50]: `channel.copyTo(output)` is outside of `launch` block and it should ensure that file should be open as long as the channel is open. The channel is closed at the end of `launch` block. Do I miss something here?

**Review by e5l — APPROVED** [review]
lgtm

**Pull request merged**